test/hotspot/jtreg/compiler/loopopts/superword/TestDependencyOffsets.java

```diff
@@ -486,15418 +409,475 @@
+                   type.name, id, type.name,
+                   type.name, id, type.name,
+                   type.name, id, type.name,
+                   id, id, id, id, secondArgument, id,
+                   // IR rules
+                   generateIRRules(),
+                   // test
+                   id, type.name, type.name,
+                   start, end,
+                   offset, type.name, loadFrom, type.operator, type.value,
+                   // run
+                   id, id, id, id, id, id, secondArgument, id, id, id, id, id, id);
+        }
+
+        /*
+         * We generate a number of IR rules for every TestDefinition. If an what kind of vectorization we
+         * expect depends on AlignVector and MaxVectorSize, as well as the byteOffset between the load and
+         * store.
+         */
+        String generateIRRules() {
+            StringBuilder builder = new StringBuilder();
+
+            // General condition for vectorization:
+            //   at least 4 bytes:    width >= 4
+            //   at least 2 elements: width >= 2 * type.size
+            int minVectorWidth = Math.max(4, 2 * type.size);
+            builder.append("    // minVectorWidth = " + minVectorWidth + " = max(4, 2 * type.size)\n");
+
+            int byteOffset = offset * type.size;
+            builder.append("    // byteOffset = " + byteOffset + " = offset * type.size\n");
+
+            // In a store-forward case, later iterations load from stores of previous iterations.
+            // If the offset is too small, that leads to cyclic dependencies in the vectors. Hence,
+            // we use shorter vectors to avoid cycles and still vectorize. Vector lengths have to
+            // be powers-of-2, and smaller or equal to the byteOffset. So we round down to the next
+            // power of two.
+            int infinity = 256; // No vector size is ever larger than this.
+            int maxVectorWidth = infinity; // no constraint by default
+            if (0 < byteOffset && byteOffset < maxVectorWidth) {
+                int log2 = 31 - Integer.numberOfLeadingZeros(offset);
+                int floorPow2 = 1 << log2;
+                maxVectorWidth = Math.min(maxVectorWidth, floorPow2 * type.size);
+                builder.append("    // Vectors must have at most " + floorPow2 +
+                               " elements: maxVectorWidth = " + maxVectorWidth +
+                               " to avoid cyclic dependency.\n");
+            }
 
-    @Run(test = "testIntM129")
-    @Warmup(0)
-    public static void runIntM129() {
-        int[] data = new int[RANGE];
-        init(data);
-        testIntM129(data);
-        verify("testIntM129", data, goldIntM129);
-    }
+            // Rule 1: No strict alignment: -XX:-AlignVector
+            IRRule r1 = new IRRule(type, type.irNode);
+            r1.addApplyIf("\"AlignVector\", \"false\"");
+            r1.addApplyIf("\"MaxVectorSize\", \">=" + minVectorWidth + "\"");
```

Comment:
Maybe at some point, the Compile Framework could offer such an IR rule builder class. But might be something for a future RFE.

Reply:
Nice idea. Probably that would rather be part of the (future) Template Framework?